qualtran/bloqs/state_preparation/controlled_state_preparation.py

```diff
@@ -0,0 +1,484 @@
+        return rom_values
+
+
+@bloq_example
+def _controlled_state_prep() -> ControlledStatePreparationUsingRotations:
+    state_coefs = (
+        (-0.42677669529663675 - 0.1767766952966366j),
+        (0.17677669529663664 - 0.4267766952966367j),
+        (0.17677669529663675 - 0.1767766952966368j),
+        (0.07322330470336305 - 0.07322330470336309j),
+        (0.4267766952966366 - 0.17677669529663692j),
+        (0.42677669529663664 + 0.17677669529663675j),
+        (0.0732233047033631 + 0.17677669529663678j),
+        (-0.07322330470336308 - 0.17677669529663678j),
+    )
+    controlled_state_prep = ControlledStatePreparationUsingRotations(
+        state_bitsizes=2, rot_reg_bitsizes=4, state=state_coefs
+    )
+    return controlled_state_prep
+
+
+_CONTROLLED_STATE_PREP_DOC = BloqDocSpec(
+    bloq_cls=ControlledStatePreparationUsingRotations,
+    import_line='from qualtran.bloqs.state_preparation.controlled_state_preparation import ControlledStatePreparationUsingRotations',
+    examples=(_controlled_state_prep,),
+)
+
+
+@attrs.frozen
+class ControlledQROMRotateQubit(Bloq):
+    r"""Array of controlled rotations $Z^{\theta_i/2}$ for a list of angles $\theta$.
+
+    It uses phase kickback and thus needs a phase gradient state in order to work. This
+    state must be provided externally for efficiency, as it is unaffected and can thus be reused.
+    Refer to [1], section on arbitrary quantum state preparation on page 3.
+
+    Args:
+        selection_bitsizes: number of qubits used for encoding the selection register of the QROM, it
+            must be equal to $\lceil \log_2(l_{rv}) \rceil$, where $l_{rv}$ is the number of angles
+            provided.
+        rot_reg_bitsizes: size of the register that is used to store the rotation angles. Bigger values
+            increase the accuracy of the results.
+        rom_values: the tuple of values to be loaded in the rom, which correspond to the angle of
+            each rotation. In order to get the rom value that corresponds to a given angle use
+            RotationTree.angle_to_rom_value(angle, rot_reg_bitsizes).
+
+    References:
+        [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+        (https://arxiv.org/abs/1812.00954).
+            Low, Kliuchnikov, Schaeffer. 2018.
+    """
+
+    selection_bitsizes: int
+    rot_reg_bitsizes: int
+    rom_values: Tuple
+
+    @property
+    def signature(self):
+        return Signature.build(
+            prepare_control=1,
+            selection=self.selection_bitsizes,
+            qubit=1,
+            rot_reg=self.rot_reg_bitsizes,
+            phase_gradient=self.rot_reg_bitsizes,
+        )
+
+    def build_composite_bloq(self, bb: BloqBuilder, **soqs: SoquetT) -> Dict[str, SoquetT]:
+        """Parameters:
+        * prepare_control
+        * selection (not necessary if selection_bitsizes == 0)
+        * qubit
+        * rot_reg
+        * phase_gradient
```

Comment:
Similarly, we can get rid of the `apply_qrom` helper and rewrite this in a more succinct way using `bb.add_d` method. 

```python
    def build_composite_bloq(self, bb: BloqBuilder, **soqs: SoquetT) -> Dict[str, SoquetT]:
        """Parameters:
        * prepare_control
        * selection (not necessary if selection_bitsizes == 0)
        * qubit
        * rot_reg
        * phase_gradient
        """
        qrom = QROM(
            [np.array(self.rom_values)],
            selection_bitsizes=(self.selection_bitsizes,),
            target_bitsizes=(self.rot_reg_bitsizes,),
            num_controls=2,
        )
        # both prepare_control and qubit will control the QROM so that this acts as a control Z
        qrom_control = bb.join(np.array([soqs.pop("prepare_control"), soqs.pop("qubit")]))
        phase_grad = soqs.pop("phase_gradient")
        # load angles in rot_reg (line 1 of eq (8) in [1])
        soqs = bb.add_d(qrom, control=qrom_control, target0_=soqs.pop("rot_reg"), **soqs)
        # phase kickback via phase_grad += rot_reg (line 2 of eq (8) in [1])
        soqs["target0_"], phase_grad = bb.add(
            AddIntoPhaseGrad(self.rot_reg_bitsizes, self.rot_reg_bitsizes),
            x=soqs["target0_"],
            phase_grad=phase_grad,
        )
        # uncompute angle load in rot_reg to disentangle it from selection register
        # (line 1 of eq (8) in [1])
        soqs = bb.add_d(qrom, **soqs)
        separated = bb.split(soqs.pop('control'))
        return {
            "prepare_control": separated[0],
            "qubit": separated[1],
            "phase_gradient": phase_grad,
            "rot_reg": soqs.pop('target0_'),
        } | soqs
```